apigee_edge.module

```diff
@@ -141,6 +149,53 @@ function apigee_edge_user_format_name_alter(&$name, AccountInterface $account) {
   }
 }
 
+/**
+ * Prepares variables for Apigee entity templates.
+ *
+ * Default template: apigee-entity.html.twig.
+ *
+ * @param array $variables
+ *   An associative array containing:
+ *   - elements: An array of elements to display in view mode.
+ */
+function template_preprocess_apigee_entity(array &$variables) {
+  $variables['view_mode'] = $variables['elements']['#view_mode'];
+  /** @var \Drupal\apigee_edge\Entity\EdgeEntityInterface $entity */
+  $entity = $variables['entity'] = $variables['elements']['#entity'];
+
+  $variables['label'] = $entity->label();
+
+  if (!$entity->isNew() && $entity->hasLinkTemplate('canonical')) {
+    $variables['url'] = $entity->toUrl('canonical', ['language' => $entity->language()])->toString();
+  }
+
+  // Helpful $content variable for templates.
+  $variables += ['content' => []];
+  foreach (Element::children($variables['elements']) as $key) {
+    $variables['content'][$key] = $variables['elements'][$key];
+  }
+}
+
+/**
+ * Implements hook_theme_suggestions_HOOK().
+ */
+function apigee_edge_theme_suggestions_apigee_entity(array $variables) {
+  $suggestions = [];
+  /** @var \Drupal\apigee_edge\Entity\EdgeEntityInterface $entity */
+  $entity = $variables['elements']['#entity'];
+  $sanitized_view_mode = str_replace('.', '_', $variables['elements']['#view_mode']);
+
+  if ($entity instanceof AppInterface) {
+    $suggestions[] = 'apigee_entity__app';
+    $suggestions[] = 'apigee_entity__app__view_mode__' . $sanitized_view_mode;
+  }
+
+  $suggestions[] = 'apigee_entity__' . $entity->getEntityTypeId();
+  $suggestions[] = 'apigee_entity__' . $entity->getEntityTypeId() . '__view_mode__' . $sanitized_view_mode;
```

Comment:
@mxr576 I'm so sorry I missed this the last time around, but the `view_mode__` part should not be here.  This is done with CSS classes, but not with theme hook suggestions (I've never seen a template with this despite the potential confusion and clashes - there is an issue for this somewhere). Could we just change these real quick?

```diff
- $suggestions[] = 'apigee_entity__app__view_mode__' . $sanitized_view_mode;
+ $suggestions[] = 'apigee_entity__app__' . $sanitized_view_mode;
```

```diff
- $suggestions[] = 'apigee_entity__' . $entity->getEntityTypeId() . '__view_mode__' . $sanitized_view_mode;
+ $suggestions[] = 'apigee_entity__' . $entity->getEntityTypeId() . '__' . $sanitized_view_mode;
```

Reply:
I implemented it like this because I agree with the answer in comment 44 to comment 43 here:
https://www.drupal.org/project/drupal/issues/2808481#comment-11784959
Although Apigee entities do not support bundles at this moment and they probably won't support them for a long time or never, but for consistency, it could make sense to use this pattern. Having "view-mode" in the template name may seem odd for the first sight, but it can not do any harm, although and unexpected, the unwaned clash could create problems. @jacine What do you think? 

This could be also related: https://www.drupal.org/project/drupal/issues/2831144

Reply:
I think it's definitely an anti-pattern. I haven't come across any templates using this pattern in over 3 years of doing D8 exclusively. Clashes are possible (we deal with this everyday in Drupal 8, think `field_image` on a node and also on a media entity), but they're not really a hinderance / a big deal. I've implemented custom theme hook suggestions hooks many times, for contextual reasons, but this was never one of them. It kinda gets into optimizing for edge case territory IMO. 

Those issues are going to take a LONG time to work out, IMO.  It's a tough issue, in general.  There are also lots of proposed changes ahead: https://www.drupal.org/project/issues/search?issue_tags=Theme%20System%20Modernization%20Initiative

So, yeah, I would take it out, but it also doesn't really do harm, so if you disagree I'm fine with you making that call.

Reply:
Ok, let's get rid of the "view mode" because that makes template suggestion names shorter and deal with the clashes if we ever start supporting bundles.